Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py

```diff
@@ -0,0 +1,1195 @@
+                area = self.area
+                old_type = area.type
+                area.ui_type = 'ShaderNodeTree'
+
+                space = area.spaces.active
+                space.node_tree = matCopy.node_tree
+
+                for node in matCopy.node_tree.nodes:
+                    if node.type == 'GROUP':
+                        node.select = True
+                        with context.temp_override(area=area):
+                            bpy.ops.node.group_ungroup()
+                        node.select = False
+
+                for node in matCopy.node_tree.nodes:
+                    node.select = False
+
+                # Restore context area type:
+                area.type = old_type
+
+                msb_log(f"Creating material copy '{mat.name}'->'{matCopy.name}'")
+
+                # Use same BSDF type if we can bake its inputs,
+                # otherwise connect the fallback baked maps to principled bsdf:
+                if canBakeBSDF:
+                    bakedBSDF = matCopy.node_tree.nodes.new(type=bsdf.bl_idname)
+                else:
+                    bakedBSDF = matCopy.node_tree.nodes.new(type='ShaderNodeBsdfPrincipled')
+
+                # Copy the input settings from the original bsdf so anything that's not baked still matches:
+                for input in bsdf.inputs:
+                    bakedBSDFInputName = self.getBSDFChannelInputName(bakedBSDF, input.name)
+                    if bakedBSDFInputName is None:
+                        continue
+
+                    bakedBSDF.inputs[input.name].default_value = input.default_value
+
+                # Find the lowest node in the tree and put the baked bsdf under that:
+                minYLocation = bsdf.location[1]
+                for node in matCopy.node_tree.nodes:
+                    minYLocation = min(minYLocation, self.getNodeYLocation(node))
+
+                bakedBSDF.location = bsdf.location  # (bsdf.location[0], minYLocation - 1000)
+                # Give bsdf a name and set its name on the material, so we can find it again:
+                bakedBSDF.name = BAKED_MATERIAL_SHADER
+                matCopy[BAKED_MATERIAL_SHADER] = bakedBSDF.name
+
+            mat = matCopy
+
+        return mat
+
+    def prepareBake(self, context, obj, bsdf, mat, canBakeBSDF):
+        if context.object is not None:
+            bpy.ops.object.mode_set(mode='OBJECT')
+        for ob in context.selected_objects:
+            ob.select_set(False)
+        obj.select_set(True)
+        bpy.context.view_layer.objects.active = obj
+
+        self.prepareObjectForBaking(context, obj)
+        mat = self.prepareMaterial(context, obj, bsdf, mat, canBakeBSDF)
+
+        return mat
+
+    def setupRenderSettings(self, context):
+        self.originalSceneSettings = []
+
+        self.setRestorableContextSetting(context, "scene.render.engine", "CYCLES")
+        self.setRestorableContextSetting(context, "scene.cycles.device", "GPU")
```

Comment:
Do we need to enforce CYCLES/GPU? Could there be a scenario where that doesn't work?

Reply:
We need cycles for baking, GPU is faster.
I've added a check now to not do this if there is no GPU.